File: Graph/FindMaxArea/max.py
from collections import deque

class Solution:
    def findTotalNodeInConnComp(self, sr, sc, image, visited):
        m = len(image)
        n = len(image[0])

        qu = deque()
        qu.append((sr, sc))
        visited[sr][sc] = True

        dx = [-1, 0, 1, 0, 1, 1, -1, -1]
        dy = [0, -1, 0, 1, 1, -1, 1, -1]

        totalNodes = 1

        while qu:
            r, c = qu.popleft()

            for i in range(8):
                nr = r + dx[i]
                nc = c + dy[i]

                if 0 <= nr < m and 0 <= nc < n:
                    if not visited[nr][nc] and image[nr][nc]:
                        qu.append((nr, nc))
                        visited[nr][nc] = True
                        totalNodes += 1

        return totalNodes

    def findMaxArea(self, grid):
        m = len(grid)
        n = len(grid[0])

        visited = [[False] * n for _ in range(m)]

        maxNodes = 0

        for i in range(m):
            for j in range(n):
                if not visited[i][j] and grid[i][j]:
                    maxNodes = max(
                        maxNodes,
                        self.findTotalNodeInConnComp(i, j, grid, visited)
                    )

        return maxNodes

File: Graph/FindMaxArea/max.cpp
#include <vector>
#include <queue>
#include <algorithm>
using namespace std;

class Solution {
public:
    int findTotalNodeInConnComp(int sr, int sc,
                                vector<vector<int>>& image,
                                vector<vector<bool>>& visited) {

        int m = image.size();
        int n = image[0].size();

        queue<vector<int>> qu;
        qu.push({sr, sc});
        visited[sr][sc] = true;

        int dx[8] = {-1, 0, 1, 0, 1, 1, -1, -1};
        int dy[8] = {0, -1, 0, 1, 1, -1, 1, -1};

        int totalNodes = 1;

        while (!qu.empty()) {
            auto curr = qu.front();
            qu.pop();

            int r = curr[0];
            int c = curr[1];

            for (int i = 0; i < 8; i++) {
                int nr = r + dx[i];
                int nc = c + dy[i];

                if (nr >= 0 && nr < m && nc >= 0 && nc < n) {
                    if (!visited[nr][nc] && image[nr][nc]) {
                        qu.push({nr, nc});
                        visited[nr][nc] = true;
                        totalNodes++;
                    }
                }
            }
        }

        return totalNodes;
    }

    int findMaxArea(vector<vector<int>>& grid) {
        int m = grid.size();
        int n = grid[0].size();

        vector<vector<bool>> visited(m, vector<bool>(n, false));

        int maxNodes = 0;

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (!visited[i][j] && grid[i][j]) {
                    maxNodes = max(maxNodes,
                                   findTotalNodeInConnComp(i, j, grid, visited));
                }
            }
        }

        return maxNodes;
    }
};